asset/encoding.go

```diff
@@ -664,3 +671,131 @@ func LeafDecoder(r io.Reader, val any, buf *[8]byte, l uint64) error {
 	}
 	return tlv.NewTypeForEncodingErr(val, "Asset")
 }
+
+func EncodeTapBranchPreimage(preimage TapBranchPreimage) []byte {
+	var encodedBranch bytes.Buffer
+	_, _ = encodedBranch.Write(preimage.left[:])
+	_, _ = encodedBranch.Write(preimage.right[:])
+
+	return encodedBranch.Bytes()
+}
+
+func DecodeTapBranchPreimage(branchData []byte) (*TapBranchPreimage, error) {
+	// Given data cannot be a valid encoded TapBranchPreimage.
+	if len(branchData) != TapBranchPreimageLen {
+		return nil, fmt.Errorf("invalid tapscript preimage length")
+	}
+
+	var leftHash, rightHash [chainhash.HashSize]byte
+	copy(leftHash[:], branchData[:chainhash.HashSize])
+	copy(rightHash[:], branchData[chainhash.HashSize:])
+
+	return &TapBranchPreimage{
+		left:  leftHash,
+		right: rightHash,
+	}, nil
+}
+
+// The following TapLeaf {en,de}coders are a duplicate of those used in
+// btcwallet. Specifically, the inner loop logic for handling []TapLeaf.
+// https://github.com/btcsuite/btcwallet/blob/master/waddrmgr/tlv.go#L160
+// This duplication is needed until we update btcwallet to export these methods.
+
+// EncodeTapLeaf encodes a TapLeaf into a byte slice containing a TapLeaf TLV
+// record, prefixed with a varint indicating the length of the record.
+func EncodeTapLeaf(leaf *txscript.TapLeaf) ([]byte, error) {
+	if leaf == nil {
+		return nil, fmt.Errorf("cannot encode nil tapleaf")
+	}
+
+	leafVersion := uint8(leaf.LeafVersion)
+	tlvRecords := []tlv.Record{
+		tlv.MakePrimitiveRecord(
+			typeTapLeafVersion, &leafVersion,
+		),
+	}
+
+	if len(leaf.Script) > 0 {
+		tlvRecords = append(
+			tlvRecords, tlv.MakePrimitiveRecord(
+				typeTapLeafScript, &leaf.Script,
+			),
+		)
+	}
+
+	tlvStream, err := tlv.NewStream(tlvRecords...)
+	if err != nil {
+		return nil, err
+	}
+
+	var leafTLVBytes bytes.Buffer
+	err = tlvStream.Encode(&leafTLVBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	// We encode the record with a varint length followed by
+	// the _raw_ TLV bytes.
+	var buf bytes.Buffer
```

Comment:
This is so it's a nested TLV? Or exterior length prefix for the DB? 

If it's a nested TLV, then the value already has a length prefix, so we can use the limit reader trick we use elsewhere.

Reply:
Yeah, I wasn't certain about carrying this bit over from the original code - there the size prefix is used to make a `limitedReader` for each `TapLeaf`, where those readers are children of a reader made with a size hint about the full `[]TapLeaf` object. For a single leaf, I don't see the utility of the size prefix.

> so we can use the limit reader trick we use elsewhere

Do you have a ref for this? Or maybe it's not needed here at all (but we should probably enforce a max tapleaf script size before encode?)

Reply:
Yeah, I also don't think we need this double encoding.
The nice thing about the TLV stream is that it handles `EOF` correctly if it's at the end of a record. So we only need to use the limit reader trick if we are reading a nested record within a larger stream. If a stream occupies the full bytes we have available, we don't need to do this.